### PR TITLE
Multi-target Json and Logger Source Generators between Roslyn v3.9 and v4.0

### PR DESCRIPTION
**Pull request by eerhardt**
Add the ability to target both Roslyn v3.9 and v4.0 in our Logging source generator. This is only enabled in the NuGet packages. For the ref-packs (both NetCoreApp and AspNetCore), we will only ship the `4.0` targeting versions of the generators. In order to target `net6.0`, customers must use VS 2022 which has Roslyn 4.0.

This is just an early draft to get initial feedback on the approach. The structure created here follows the proposal in https://github.com/dotnet/sdk/issues/20355.

If the SDK proposal isn't accepted for 6.0, my plan is to ship `.targets` files in both NuGet packages which will select the correct source generator version using custom logic. In both cases, we will need these changes to build the source generator twice - once for each Roslyn version we want to target.

Once we settle on an approach here, I can make the same changes to the System.Text.Json source generator.

### CODE REVIEW

**Review by ViktorHofer — APPROVED** [review]
The msbuild and the project changes as well as the transport package change LGTM.

**Comment by eerhardt** [comment]
@ViktorHofer - one place I was wondering about was the changes to the `.sln` file. I know we generate our .sln files. Will the sln generator handle these changes?

File: src/libraries/Directory.Build.targets
Comment (eerhardt) [reply src/libraries/Directory.Build.targets:263]: Right now: nothing. But check out https://github.com/dotnet/sdk/issues/20355 for more information.

**Comment by ViktorHofer** [comment]
> @ViktorHofer - one place I was wondering about was the changes to the .sln file. I know we generate our .sln files. Will the sln generator handle these changes?

Feel free to check the solution file format change in as it probably will be a some time until we regenerate them again. Regarding your question, yes slngen discovers these references as [`AnalyzerReference` items are just slightly modified ProjectReferences](https://github.com/dotnet/runtime/blob/933988c35c172068652162adf6f20477231f815e/src/libraries/Directory.Build.targets#L244-L256). Because of that, the analyzers are part of the depedency graph and slngen sees them.

File: eng/MultiTargetRoslynComponent.targets.template
Comment (ericstj) [inline eng/MultiTargetRoslynComponent.targets.template:20]: Why didn't you just use `{TargetPrefix}` to give this item a namespace?

File: eng/MultiTargetRoslynComponent.targets.template
Comment (ericstj) [inline eng/MultiTargetRoslynComponent.targets.template:16]: This might be too late.  Wouldn't conflict resolution run before this and drop one of them?  I guess that would depend on if we gave them the same name or not.  Actually maybe we need to give different names per @sharwell's sxs versioning suggestion.

Comment (eerhardt) [reply eng/MultiTargetRoslynComponent.targets.template:16]: In my testing, it isn't too late. `_HandlePackageFileConflicts` outputs:

`Encountered conflict between 'Analyzer:C:\Users\eerhardt\.nuget\packages\system.text.json\7.0.0-dev\analyzers\dotnet\roslyn3.9\cs\System.Text.Json.SourceGeneration.dll' and 'Analyzer:C:\Users\eerhardt\.nuget\packages\system.text.json\7.0.0-dev\analyzers\dotnet\roslyn4.0\cs\System.Text.Json.SourceGeneration.dll'. Could not determine winner due to equal file and assembly versions.`

So both Analyzers are still around during `ResolveAssemblyReferences`.

However, I can move this up to be `AfterTargets="ResolvePackageDependenciesForBuild"`. Or potentially even `ResolveLockFileAnalyzers`. Thoughts?

**Comment by eerhardt** [comment]
I believe this is ready for review, assuming https://github.com/dotnet/sdk/pull/20793 gets approved and merged for 6.0.

File: eng/MultiTargetRoslynComponent.targets.template
Comment (ericstj) [reply eng/MultiTargetRoslynComponent.targets.template]: We need your logic to run before conflict resolution.  The fact that you are seeing `Encountered conflict ` means it's not early enough.  Your logic needs to behave "as if" it were NuGet making the decision.  `AfterTargets` the target(s) that raise(s) these analyzers from NuGet packages would be best if it is / they are public.

Comment (eerhardt) [reply eng/MultiTargetRoslynComponent.targets.template]: Both `ResolvePackageDependenciesForBuild` and `ResolveLockFileAnalyzers` happen before conflict resolution runs. And `ResolvePackageDependenciesForBuild` depends on `ResolveLockFileAnalyzers`.

The flow, as I understand it is:

1. [ResolvePackageAssets](https://github.com/dotnet/sdk/blob/17181d34b143cbf4b957abdaabb1d2fb116dc612/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets#L246-L275) outputs the analyzers in the project.assets.json to the `@(ResolvedAnalyzers)` item.
2. [ResolveLockFileAnalyzers](https://github.com/dotnet/sdk/blob/17181d34b143cbf4b957abdaabb1d2fb116dc612/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets#L485-L490) depends on `ResolvePackageAssets` and turns `@(ResolvedAnalyzers)` into `@(Analyzer)`
3. [ResolvePackageDependenciesForBuild](https://github.com/dotnet/sdk/blob/17181d34b143cbf4b957abdaabb1d2fb116dc612/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets#L135-L149) depends on `ResolveLockFileAnalyzers`, and doesn't have the "LockFile" in the name.

Another thing occurred to me as I wrote this - none of these targets occur with a non-SDK .csproj that uses `PackageReference`. Do you think that is a scenario we need to support?

Comment (ericstj) [reply eng/MultiTargetRoslynComponent.targets.template]: > none of these targets occur with a non-SDK .csproj that uses PackageReference. Do you think that is a scenario we need to support?

I believe that scenario would still raise analyzers and run in VS 2022, so yes.  Our packages support install in classic .NETFramework projects.  The nice thing about `AfterTargets` is that it's a noop if the target doesn't exist, and it'll only run once for a given project/global-property-set.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.3.9.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.3.9.cs:1]: Nit: filename seems a bit odd to me. How about something like `JsonSouceGenerator.Roslyn39.cs`?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.3.9.cs:1]: or perhaps `RoslynV3` and `RoslynV4`.

File: eng/MultiTargetRoslynComponent.targets.template
Comment (eerhardt) [reply eng/MultiTargetRoslynComponent.targets.template]: Fixed.

File: eng/MultiTargetRoslynComponent.targets.template
Comment (eerhardt) [reply eng/MultiTargetRoslynComponent.targets.template]: Thanks for the help. I changed this target to run `AfterTargets="ResolvePackageDependenciesForBuild;ResolveNuGetPackageAssets"`, which should address these concerns.

File: src/libraries/System.Text.Json/gen/Reflection/TypeWrapper.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/Reflection/TypeWrapper.cs:317]: I wish could avoid this. Effectively it will break VS19 sourcegen for any type containing value tuples. Would it be possible to consider potential workarounds for 3.9 or upgrade to 3.11?

One potential workaround would be to check if `fieldSymbol.Name` does not match the `Item[1-7]`.

File: src/libraries/System.Text.Json/src/System.Text.Json.csproj
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System.Text.Json.csproj:352]: Similarly, consider using a `.Roslyn3.csproj` naming convention here.

File: src/libraries/System.Text.Json/gen/Reflection/TypeWrapper.cs
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/Reflection/TypeWrapper.cs:317]: This might also provide hints on how the property could be derived:

https://github.com/dotnet/roslyn/blob/315c2e149ba7889b0937d872274c33fcbfe9af5f/src/Compilers/CSharp/Portable/Symbols/Tuples/TupleErrorFieldSymbol.cs#L94

Comment (eerhardt) [reply src/libraries/System.Text.Json/gen/Reflection/TypeWrapper.cs:317]: @jcouv @chsienki - any thoughts here? Is there a way we can simulate the `fieldSymbol.IsExplicitlyNamedTupleElement` logic while targeting Roslyn 3.9? Or should we just target 3.10 which has this API?

Comment (eerhardt) [reply src/libraries/System.Text.Json/gen/Reflection/TypeWrapper.cs:317]: In an offline discussion with @jaredpar, I decided to upgrade to 3.11. If anyone has concerns/feedback, we can revisit this decision.

File: src/libraries/System.Text.Json/src/System.Text.Json.csproj
Comment (layomia) [reply src/libraries/System.Text.Json/src/System.Text.Json.csproj:352]: +1. Or `RoslynV3` as suggested elsewhere so it's clearer what the numbers mean.

Comment (eerhardt) [reply src/libraries/System.Text.Json/src/System.Text.Json.csproj]: I chose to go with:
* System.Text.Json.SourceGeneration.Roslyn3.11.csproj
* System.Text.Json.SourceGeneration.Roslyn4.0.csproj
* System.Text.Json.SourceGeneration.Roslyn3.11.Unit.Tests.csproj
* etc

**Review by layomia — APPROVED** [review]
LGTM assuming clean CI & there are no issues relating to https://github.com/dotnet/sdk/pull/20793.

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/Microsoft.Extensions.Logging.Abstractions.sln
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/Microsoft.Extensions.Logging.Abstractions.sln:20]: so with future upgrades we have to rename these files?

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:440]: I think the latest version is using `BaseNamespaceDeclarationSyntax`

see https://github.com/dotnet/runtime/pull/57894/files

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/Microsoft.Extensions.Logging.Abstractions.sln
Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/Microsoft.Extensions.Logging.Abstractions.sln:20]: if yes, do we prefer to use an intermediary name that doesnt have to change with version future changes?

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/Microsoft.Extensions.Logging.Abstractions.sln:20]: My thinking is that future upgrades would mean a new file with the appropriate version in the name. If we were no longer targeting a specific version, we would then delete the file for that version. So if were no longer targeting Roslyn v4.0, but instead wanted to target Roslyn v4.1, we would delete the 4.0 file and add a 4.1 file.

Do you have an idea on what an intermediary name pattern would look like that would make sense, but not require the version number in it?

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:440]: These changes are sync'd to that change. Note that in the #57894 change, line 447 uses `BaseNamespaceDeclarationSyntax`, but lines 442-443 use `NamespaceDeclarationSyntax` and `FileScopedNamespaceDeclarationSyntax`, as this change does.

**Comment by eerhardt** [comment]
Windows failures are #58927.

Linux_musl x64 failures are infrastructure related.

This has passed CI previously before resolving merge conflicts. Libraries legs are building and tests are passing. Merging to unblock 6.0 work.

**Pull request merged**
